File: python/be/api/venues/kraken.py
"""Read-only client for Kraken's public API. Mirror of node/be/api/venues/kraken.js.

Public market data only (Assets, AssetPairs, Ticker), no key. A transport
failure, a non-JSON body, or a non-empty Kraken `error` array all mean the
figure cannot be observed -- SiteUnreachable, which grades Blocked, never
Failed. The live exchange is not a system under test.
"""

from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request

BASE = os.environ.get("KRAKEN_URL", "https://api.kraken.com/0/public").rstrip("/")
TIMEOUT_S = 15
USER_AGENT = "qa-automation-crypto-cex/1.0 (read-only invariants)"


class SiteUnreachable(Exception):
    pass


def _call(path_and_query):
    req = urllib.request.Request(BASE + path_and_query, headers={"User-Agent": USER_AGENT, "Accept": "application/json"})
    try:
        with urllib.request.urlopen(req, timeout=TIMEOUT_S) as res:
            status, text = res.status, res.read().decode("utf-8", "replace")
    except urllib.error.HTTPError as err:
        if err.code in (403, 429) or err.code >= 500:
            raise SiteUnreachable(f"Kraken returned HTTP {err.code} for {path_and_query}") from err
        status, text = err.code, err.read().decode("utf-8", "replace")
    except (urllib.error.URLError, TimeoutError, OSError) as err:
        raise SiteUnreachable(f"Kraken did not answer {path_and_query}: {err}") from err
    try:
        body = json.loads(text) if text else None
    except json.JSONDecodeError:
        body = None
    if not isinstance(body, dict):
        raise SiteUnreachable(f"Kraken served a non-JSON body for {path_and_query} (a challenge or error page)")
    if isinstance(body.get("error"), list) and body["error"]:
        raise SiteUnreachable("Kraken error for " + path_and_query + ": " + ", ".join(body["error"]))
    if not isinstance(body.get("result"), dict):
        raise SiteUnreachable(f"Kraken returned no result for {path_and_query}")
    return body["result"]


class Kraken:
    SiteUnreachable = SiteUnreachable

    @staticmethod
    def assets():
        return _call("/Assets")

    @staticmethod
    def asset_pairs():
        return _call("/AssetPairs")

    @staticmethod
    def ticker(pair):
        return _call("/Ticker?pair=" + urllib.parse.quote(pair))

    @staticmethod
    def pair_names():
        r = Kraken.asset_pairs()
        return [p.get("wsname") for p in r.values() if p.get("wsname")]

    @staticmethod
    def last_price(pair):
        r = Kraken.ticker(pair)
        row = next(iter(r.values()), None)
        if not row or not isinstance(row.get("c"), list) or not row["c"]:
            raise SiteUnreachable(f"Kraken ticker for {pair} had no last price")
        n = float(row["c"][0])
        if not (n > 0):
            raise SiteUnreachable(f"Kraken ticker for {pair} had a non-numeric price")
        return n

File: node/be/api/venues/kraken.js
'use strict';

/**
 * Read-only client for Kraken's public API -- the live exchange this domain
 * cross-checks the mini-cex against. No key, no account: only public market
 * data (Assets, AssetPairs, Ticker).
 *
 * Kraken wraps every response in { error: [...], result: {...} }. A transport
 * failure, a non-JSON body (a challenge or error page), or a non-empty `error`
 * array all mean the figure cannot be observed -- SiteUnreachable, which grades
 * Blocked, never Failed. The live exchange is not a system under test.
 */

const BASE = (process.env.KRAKEN_URL || 'https://api.kraken.com/0/public').replace(/\/+$/, '');
const TIMEOUT_MS = 15000;
const USER_AGENT = 'qa-automation-crypto-cex/1.0 (read-only invariants)';

class SiteUnreachable extends Error {}

async function call(pathAndQuery) {
  let res;
  try {
    res = await fetch(BASE + pathAndQuery, { headers: { 'user-agent': USER_AGENT, accept: 'application/json' }, signal: AbortSignal.timeout(TIMEOUT_MS) });
  } catch (err) {
    throw new SiteUnreachable('Kraken did not answer ' + pathAndQuery + ': ' + (err.cause && err.cause.message ? err.cause.message : err.message));
  }
  if (res.status === 403 || res.status === 429 || res.status >= 500) throw new SiteUnreachable('Kraken returned HTTP ' + res.status + ' for ' + pathAndQuery);
  const text = await res.text();
  let body = null; try { body = text ? JSON.parse(text) : null; } catch { body = null; }
  if (body === null || typeof body !== 'object') throw new SiteUnreachable('Kraken served a non-JSON body for ' + pathAndQuery + ' (a challenge or error page)');
  if (Array.isArray(body.error) && body.error.length) throw new SiteUnreachable('Kraken error for ' + pathAndQuery + ': ' + body.error.join(', '));
  if (!body.result || typeof body.result !== 'object') throw new SiteUnreachable('Kraken returned no result for ' + pathAndQuery);
  return body.result;
}

const Kraken = {
  SiteUnreachable,
  assets() { return call('/Assets'); },
  assetPairs() { return call('/AssetPairs'); },
  ticker(pair) { return call('/Ticker?pair=' + encodeURIComponent(pair)); },

  /** The set of wsname strings (e.g. "XBT/USD") the exchange lists. */
  async pairNames() {
    const r = await this.assetPairs();
    return Object.values(r).map((p) => p.wsname).filter(Boolean);
  },
  /** The last-trade price of a pair as a number, from ticker.c[0]. */
  async lastPrice(pair) {
    const r = await this.ticker(pair);
    const row = Object.values(r)[0];
    if (!row || !Array.isArray(row.c) || row.c.length === 0) throw new SiteUnreachable('Kraken ticker for ' + pair + ' had no last price');
    const n = Number(row.c[0]);
    if (!Number.isFinite(n) || n <= 0) throw new SiteUnreachable('Kraken ticker for ' + pair + ' had a non-numeric price');
    return n;
  },
};

module.exports = { Kraken, SiteUnreachable, BASE };
